### PR TITLE
Add kms key

### PR DESCRIPTION
**Pull request by tthyer**
As previously discussed, I'm adding a KMS key for use in encoding/decoding ssm keys, based on the pattern found in aws-infra/essentials. This has been tested in sandbox with the admin role listed as one of the principles. I have removed that role for used in scicomp.

### CODE REVIEW

File: templates/toil-infra-essentials.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            Principal:
+              AWS:
+                - !Join
+                  - ''
+                  - - 'arn:aws:iam::'
+                    - !Ref AWS::AccountId
+                    - ':root'
+                - !ImportValue us-east-1-bootstrap-TravisUserArn
+                - !ImportValue us-east-1-bootstrap-CfServiceRoleArn
+                - !ImportValue us-east-1-rna-seq-reprocessing-instance-role-v001-ToilClusterRole
```
Comment (zaro0508) [inline templates/toil-infra-essentials.yaml]: this section grants permission to administer the kms key.  i think you just need to give toil cluster permission to use the key which is done below.

File: templates/toil-infra-essentials.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+              - "kms:CancelKeyDeletion"
+            Resource: "*"
+          -
+            Sid: "Allow use of the key"
+            Effect: "Allow"
+            Principal:
+              AWS:
+                - !ImportValue us-east-1-bootstrap-TravisUserArn
+                - !ImportValue us-east-1-bootstrap-CfServiceRoleArn
+                - !ImportValue us-east-1-rna-seq-reprocessing-instance-role-v001-ToilClusterRole
```
Comment (zaro0508) [inline templates/toil-infra-essentials.yaml]: gives permission to use the key so you can remove the same line in the section above.

File: templates/toil-infra-essentials.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+              - "kms:Encrypt"
+              - "kms:Decrypt"
+              - "kms:ReEncrypt*"
+              - "kms:GenerateDataKey*"
+              - "kms:DescribeKey"
+            Resource: "*"
+  AWSKmsInfraKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      AliasName: !Join
```
Comment (zaro0508) [inline templates/toil-infra-essentials.yaml]: this alias already exists, you will want to give it a different name.

File: templates/toil-infra-essentials.yaml (hunk 2 above)
Comment (zaro0508) [reply templates/toil-infra-essentials.yaml]: Also this depends on the the output in `templates/toil-cluster-role-and-policy.yaml`.  Sceptre does not guarantee running templates in a certain order.  You'll need to split this PR into two PRs and manually synchronize merging them in the correct order.

Comment (zaro0508) [reply templates/toil-infra-essentials.yaml]: Sorry, i take that back.  i forgot that you can synchroize the deploy order with [sceptre dependencies](https://sceptre.cloudreach.com/latest/docs/stack_config.html#dependencies).  An example of that definition is here: https://github.com/Sage-Bionetworks/sandbox-infra/blob/master/config/prod/accounts.yaml#L3

Comment (tthyer) [reply templates/toil-infra-essentials.yaml]: @zaro0508 in your comment above, "gives permission to use the key..." is the subject missing? the instance role does need to be able to use the key

Comment (zaro0508) [reply templates/toil-infra-essentials.yaml]: i just mean that setting principal here will give permission to use the kms key which is you need.

File: templates/toil-infra-essentials.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            Resource: "*"
+  AWSKmsToilInfraKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      AliasName: !Join
+        - ''
+        - - 'alias/'
+          - !Ref AWS::StackName
+          - '/InfraKey'
+      TargetKeyId: !Ref AWSKmsToilInfraKeyAlias
```
Comment (zaro0508) [inline templates/toil-infra-essentials.yaml]: this should be `!Ref AWSKmsInfraKey`

**Comment by tthyer** [comment]
@zaro0508 the lint failure is on circular dependencies, which is arising I b/c I used the sceptre dependencies statement. I think I'm just going to divide the role export into a different PR.

**Pull request merged**